src/main/java/com/redhat/devtools/lsp4ij/features/documentation/MarkdownConverter.java

```diff
@@ -91,28 +101,56 @@ public NodeRenderer apply(DataHolder options) {
      *                 the syntax coloration to use if MarkDown content contains some code block or blockquote to highlight.
      * @return the given <code>markdown</code> to Html.
      */
-    public @NotNull String toHtml(@NotNull String markdown, @Nullable PsiFile file) {
-        return toHtml(markdown, file != null ? file.getLanguage() : null, file != null ? file.getName() : null);
+    public @NotNull String toHtml(@NotNull String markdown,
+                                  @Nullable PsiFile file) {
+        var htmlRenderer = this.htmlRenderer;
+        if (file != null) {
+            htmlRenderer = file.getUserData(HTML_RENDERER_KEY);
+            if (htmlRenderer == null) {
+                htmlRenderer = getHtmlRenderer(file);
+            }
+        }
+        return htmlRenderer.render(htmlParser.parse(markdown));
+    }
+
+    private synchronized HtmlRenderer getHtmlRenderer(PsiFile file) {
+        var htmlRenderer = file.getUserData(HTML_RENDERER_KEY);
+        if (htmlRenderer != null) {
+            return htmlRenderer;
+        }
+
+        MutableDataSet fileOptions = new MutableDataSet(options);
+        fileOptions.set(LANGUAGE_CONTEXT, file.getLanguage());
+        fileOptions.set(FILE_NAME_CONTEXT, file.getName());
+        Path baseDir = file.getVirtualFile().getParent().getFileSystem().getNioPath(file.getVirtualFile().getParent());
+        fileOptions.set(BASE_DIR_CONTEXT, baseDir);
+
+        htmlRenderer = createHtmlRenderer(fileOptions);
+        file.putUserData(HTML_RENDERER_KEY, htmlRenderer);
```

Comment:
when is this released for GC? Are closed files user data purged when closed?

Reply:
I have changed my code and now I store HtmlRenderer in the LSPFileSupport which is eveicted when file is closed.